### DISCUSSION TITLE
I absolutely love the new direction of SvelteKit

### DISCUSSION THREAD

**Discussion by AlbertMarashi** [sveltejs/kit#6062] [General]
Don't have much time to type this, but I'm really happy with all the recent changes SvelteKit team has done.

The new load api, filesystem routing and other changes are great. People will come around despite how annoying some of the changes may seem.

After using it for some time, it's great, but the documentation was kind of difficult to understand and limited.

The new way of managing types and the automatic type creation with SvelteKit is great.

**Comment by rmunn** [comment]
Between this and https://github.com/sveltejs/kit/discussions/6005, the upvotes might turn into an informal poll.

**Reply by oxisto** [reply]
Which BTW I never intended to :) I just wanted to relieve some of my emotional pain, so did others apparently by the various responses :)

**Comment by AlbertMarashi** [comment]
Docs definitely were a bit confusing, found it difficult as hell to migrate and also confused about how everything worked, but I eventually figured it all out

I understand this is all WIP so i'm not complaining

**Reply by dummdidumm** [reply]
Anything specific in the docs or the migration we could do better?

**Reply by x-GF-x** [reply]
I needed to turn to the RealWorld example to figure out the best way to organize my routes... In my case, when returning data from PUT/POST. Initially after migration, I had a route [id]/+page.server.js, and this file contained both load and PUT. It was not intuitive to me where to move my PUT call in order to return data, but following the RW example, I now have two folders- [id], and [id].json, with +server.js being in the second folder and containing my PUT call. (If I am still missing the mark, please correct me!) But nowhere in the docs did I see a mention of using .json in the routing names and I was scratching my head for a while about best practices.

**Reply by AlbertMarashi** [reply]
> Anything specific in the docs or the migration we could do better?

Some TypeScript examples of the `import("./$types").LayoutLoad` and etc would've been useful.

**Reply by oxisto** [reply]
> > Anything specific in the docs or the migration we could do better?
> 
> Some TypeScript examples of the `import("./$types").LayoutLoad` and etc would've been useful.

Agree. In general, the docs do not really focus on typescript too much. It actually took me some googling to find out that you can type guard your `load` in TypeScript using
```ts
import type { LoadEvent } from '@sveltejs/kit';
export async function load({ fetch, params }: LoadEvent) { }
```

At least I didn't see that in the docs.

**Reply by AlbertMarashi** [reply]
@oxisto  actually I think the proper way is to use the non-generic form of LoadEvent, eg:

```ts
export const load: import("./$types").PageLoad = async () => {
   ...
}
```

This lets you get proper typing from parent functions and the layout data loading

**Reply by jr-trinium** [reply]
I'm still working on learning SK along with all the recent changes, but what about:

```javascript
// +page.ts
import type { PageLoad, PageLoadEvent } from './$types'

export const load: PageLoad = async (event: PageLoadEvent) => {
  ...
}
```

or destructured:

```javascript
// +page.ts
import type { PageLoad, PageLoadEvent } from './$types'

export const load: PageLoad = async ({ data, depends, fetch, params, parent, routeId, setHeaders, url }: PageLoadEvent) => {
  ...
}
```

The same for the other files:

```javascript
// +page.server.ts:
import type { PageServerLoad, PageServerLoadEvent } from './$types'

export const load: PageServerLoad = async (event: PageServerLoadEvent) => { ... }
export const load: PageServerLoad = async ({ clientAddress, locals, params, parent, platform, request, routeId, setHeaders, url }: PageServerLoadEvent) => { ... }
```

```javascript
// +layout.ts:
import type { LayoutLoad, LayoutLoadEvent } from './$types'

export const load: LayoutLoad = async (event: LayoutLoadEvent) => { ... }
export const load: LayoutLoad = async ({ data, depends, fetch, params, parent, routeId, setHeaders, url }: LayoutLoadEvent) => { ... }
```

```javascript
// +layout.server.ts:
import type { LayoutServerLoad, LayoutServerLoadEvent } from './$types'

export const load: LayoutServerLoad = async (event: LayoutServerLoadEvent) => { ... }
export const load: LayoutServerLoad = async ({ clientAddress, locals, params, parent, platform, request, routeId, setHeaders, url }: LayoutServerLoadEvent) => { ... }
```

Technically, you don't need to type the event since typing the load function already implies it.

**Reply by AlbertMarashi** [reply]
> 

Yes, the event parameter does not need to be typed, as the load function is already typed. These examples should be useful to others

**Reply by danawoodman** [reply]
Should be able to just do this:

```typescript
// +page.ts
import type { PageLoad } from './$types'

export const load: PageLoad = async (event) => {
  ...
}
```

No need to add the parameter typing; it's inferred.

**Reply by samuelstroschein** [reply]
The most "TS" way in my opinion: 

```ts
import type { PageLoadEvent } from './$types';

export function load(event: PageLoadEvent) {
	return {
		example: 'hello'
	};
}
```

- Make sure that the dev server is running. Otherwise, the types are not generated.

**Reply by danawoodman** [reply]
Why is this the most "TS" way? Arrow or named functions are both just ES6

This doesn't type the return value unless I'm mistaken.

**Reply by samuelstroschein** [reply]
Most "TS", or rather JS, way because it's a function. Explaining `const load = async (event) => {` to beginners is challenging. 

> This doesn't type the return value unless I'm mistaken.

That is true but because it's not necessary and was annoying. The return type is generated by SvelteKit based on the function. Aka no matter what the return type is, it will change the moment the return statement is changed.

**Reply by danawoodman** [reply]
I also prefer `function` to define named functions but there is nothing wrong with someone using/preferring arrow functions instead.

**Comment by ZetiMente** [comment]
Umm, the Routing page still isn't updated...

![Routing](https://user-images.githubusercontent.com/76985631/185997285-8573324d-c33c-4eed-b26d-4b32b1f4f209.png)

**Reply by dummdidumm** [reply]
What's incorrect about it in your opinion?

**Reply by ZetiMente** [reply]
src/routes/about does not create an /about route? src/routes/about/+page.svelte create an /about route ?

**Reply by eddiemcconkie** [reply]
Routes are not the same as pages. A route can lead to a page or an endpoint depending on the route file inside the directory, so the description is accurate

**Reply by ZetiMente** [reply]
The description is accurate to you because you are filing in context that a beginner wouldn't have. 

For starters, src/routes/about could be a file or a folder. Something like `src/routes/about/` would be much more clear since people might have the impression that they can still do about.svelte. (You have to remember, all the Youtube videos, blogs and educational material is still on the old routing system) 

Secondly, a file inside the directory actually isn't a route, it's a component because everything is private by default unless it has a + at the front of it. Therefore only `src/routes/about/+` would be true. 

Third, only pages and endpoints of the exact names of `+page.svelte`, `+page.js` or `+page.server.js` are actually routes. So just any file in the directory wouldn't work. So it's not accurate or at least clear enough to anyone not familiar with SK. 

Finally, There is no clarity at all that the route only exist if it has + in front of the name of the file. That should be in big cap letters. ONLY FILES OF THESE NAMES WILL WORK. A FILE MUST HAVE A + IN FRONT OF IT TO WORK.

The documentation before jumped into the differences of a page and endpoints. Now with a more confusing routing system, it isn't clear at all that the + makes a route public and that the name of `+page.svelte`/`+page.js`/`+page.server.js` is required. The docs give the impression that the `+page.svelte` could be anything like `+help.about` or `+about.svelte`.

Like if you were reading that for the first time, you would think they were using `+page.svelte` as an example. Not as that it actually has to be exact `+page.svelte`. You may even  do `about.svelte` because you probably wouldn't have grasped that + makes it public from only "Each route directory contains one or more route files, which can be identified by their + prefix." 

If the idea now is that people see the +, get confused and come to RTFM. Shouldn't the docs be extremely explicit on what is happening?

**Reply by braebo** [reply]
@ZetiMente Everything is explained right below your screenshot.

<img width="861" alt="Screenshot 2022-08-22 at 9 22 02 PM" src="https://user-images.githubusercontent.com/55504972/186047562-3abb44cf-3563-4d03-93f8-8cbe1333595e.png">

**Reply by ZetiMente** [reply]
How is that everything explained? That sentence is very vague and It doesn't make clear at all that the + is required. Nor does it make it clear that the page name has to be  `+page.svelte`. Someone could think that page is a placeholder for any name.

**Reply by oxisto** [reply]
> What's incorrect about it in your opinion?

It could be missunderstand that the mere existence of the directory already creates the route, which is incorrect. If one creates an empty `src/routes/about` folder, no route is created. Only by putting *at least*. one route file in there, the route is actually created.
